File: src/api/analytics.ts
import api from "./client";

export interface MetricComponent {
  label: string;
  value: number | null;
  unit?: string | null;
  extra?: Record<string, unknown> | null;
}

export interface MetricScore {
  id: string;
  name: string;
  category: string;
  description: string;
  direction: "higher_is_better" | "lower_is_better" | "mixed";
  value: number | null;
  unit?: string | null;
  components: MetricComponent[];
  tags: string[];
}

export interface AthleteMetricsResponse {
  athlete_id: number;
  metrics: MetricScore[];
}

export interface RankingEntry {
  athlete_id: number;
  full_name: string;
  value: number;
  unit?: string | null;
  team?: string | null;
  age?: number | null;
  gender?: string | null;
}

export interface MetricRankingResponse {
  metric: MetricScore;
  entries: RankingEntry[];
}

export interface MetricRankingParams {
  limit?: number;
  gender?: string;
  age_category?: string;
  client_id?: number;
}

export const fetchAthleteMetrics = async (
  athleteId: number,
  metrics?: string[]
) => {
  const params = metrics?.length ? { metric: metrics } : undefined;
  const { data } = await api.get<AthleteMetricsResponse>(
    `/analytics/athletes/${athleteId}/metrics`,
    { params }
  );
  return data;
};

export const fetchMetricRanking = async (
  metricId: string,
  params: MetricRankingParams
) => {
  const { data } = await api.get<MetricRankingResponse>(
    `/analytics/rankings/metrics/${metricId}`,
    { params }
  );
  return data;
};

File: backend/app/schemas/analytics.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel


class MetricComponent(BaseModel):
    label: str
    value: float | None
    unit: str | None = None
    extra: dict[str, Any] | None = None


class MetricScore(BaseModel):
    id: str
    name: str
    category: str
    description: str
    direction: Literal["higher_is_better", "lower_is_better", "mixed"]
    value: float | None
    unit: str | None = None
    components: list[MetricComponent] = []
    tags: list[str] = []


class AthleteMetricsResponse(BaseModel):
    athlete_id: int
    metrics: list[MetricScore]


class RankingEntry(BaseModel):
    athlete_id: int
    full_name: str
    value: float
    unit: str | None = None
    team: str | None = None
    age: int | None = None
    gender: str | None = None


class MetricRankingResponse(BaseModel):
    metric: MetricScore
    entries: list[RankingEntry]


class GameLeaderboardEntry(BaseModel):
    athlete_id: int
    full_name: str
    value: float
    team: str | None = None
    matches: int | None = None


class GameLeaderboardResponse(BaseModel):
    leaderboard_type: Literal["top_scorers", "top_shooters"]
    entries: list[GameLeaderboardEntry]


class LeaderboardEntry(BaseModel):
    athlete_id: int
    full_name: str
    team: str | None = None
    age_category: str | None = None
    position: str | None = None
    goals: int = 0
    shootout_goals: int = 0


class LeaderboardResponse(BaseModel):
    leaderboard_type: Literal["scorers", "shootouts"]
    entries: list[LeaderboardEntry]
